### PR TITLE
Schedule performance tests differently

### PR DESCRIPTION
**Pull request by wolfs**
Instead of using a task which calls a single build on CI with different parameters for distributing performance tests, we split all the performance tests into a fixed number of buckets and run them per bucket. This is the same setup we have for running our functional tests, adopted to performance tests.

In order to do the above, we create a task which runs all scenarios for a given test project. By doing so, the test project does not have to be part of the parameterization of the scenario itself (no where clause). Then, we create a JSON file which details which scenario should run for which test project on which operating system at which stage. The Teamcity DSL uses this JSON file and the timings of the tests from another file to split the scenarios into different buckets for the test projects.

I created an [Experiment Performance project](https://builds.gradle.org/project/Gradle_Check_ExperimentalPerformanceLinuxPerformanceTest?focusLine=0&mode=builds) in Teamcity, which contains the builds for splitting the performance tests into 40 buckets. The goal of this PR is to run everything we run as part of the `Performance test coordinator` in Ready for Merge in these buckets.

As a first step, I converted `JavaUpToDatePerformanceTest` to the new infrastructure and split it amongst the buckets. Since we only run 3 scenarios in that test, 37 buckets remain empty.

These are the follow up steps:
- [x] Convert more/all performance tests to not include the test project
- [x] Add all performance tests for the coordinator in Ready for Merge to `performance-tests-ci.json` and read that file to determine which scenarios to run.
- [x] Add the timings from the performance test database to `performance-test-runtimes.csv` and read that file to determine how to split the scenarios into buckets.
- [ ] Add a task to automatically update `performance-test-runtimes.csv`
- [x] Take the test project into account when looking for the history of a scenario and update the scenario names in the performance test database
- [ ] Make sure performance test tasks and template tasks do not run in parallel in one build.
- [x] Come up with a solution how to run the flakiness detection performance tests
- [x] Decide if and how to create a report containing all the performance tests for a stage.
- [x] Move the Gradle profiler html/csv reports to a directory per test.

Fixes gradle/gradle-private#3196

### CODE REVIEW

**Comment by wolfs** [comment]
Here is the successful run of one of the buckets containing one of the scenarios from `JavaUpToDatePerformanceTest`: https://builds.gradle.org/buildConfiguration/Gradle_Check_ExperimentalPerformanceLinuxPerformanceTestbucket1/38085322?focusLine=0

File: .teamcity/Gradle_Check/model/FunctionalTestBucketProvider.kt [hunk 1]
```diff
@@ -220,7 +180,7 @@ class GradleVersionRangeCrossVersionTestBucket(private val startInclusive: Strin
         )
 }
 
-class LargeSubprojectSplitBucket(val subproject: GradleSubproject, val number: Int, val include: Boolean, val classes: List<TestClassTime>) : BuildTypeBucket by subproject {
+class LargeSubprojectSplitBucket(val subproject: GradleSubproject, number: Int, val include: Boolean, val classes: List<TestClassTime>) : BuildTypeBucket by subproject {
```
Comment (lptr) [inline .teamcity/Gradle_Check/model/FunctionalTestBucketProvider.kt:183]: Why this change?

File: .teamcity/Gradle_Check/model/bucket-extensions.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * Split a list of elements into nearly even sublist. If an element is too large, largeElementSplitFunction will be used to split the large element into several smaller pieces;
+ * if some elements are too small, they will be aggregated by smallElementAggregateFunction.
+ *
+ * @param list the list to split, must be ordered by size desc
+ * @param toIntFunction the function used to map the element to its "size"
+ * @param largeElementSplitFunction the function used to further split the large element into smaller pieces
+ * @param smallElementAggregateFunction the function used to aggregate tiny elements into a large bucket
+ * @param expectedBucketNumber the return value's size should be expectedBucketNumber
+ */
+fun <T, R> split(list: LinkedList<T>, toIntFunction: (T) -> Int, largeElementSplitFunction: (T, Int) -> List<R>, smallElementAggregateFunction: (List<T>) -> R, expectedBucketNumber: Int, maxNumberInBucket: Int, noElementSplitFunction: (Int) -> List<R> = { throw IllegalArgumentException("More buckets than things to split") }): List<R> {
```
Comment (lptr) [inline .teamcity/Gradle_Check/model/bucket-extensions.kt:31]: Can we add a more descriptive name to this method? Also, let's split the parameters to separate lines.

File: buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy [hunk 3]
```diff
@@ -227,6 +271,7 @@ abstract class PerformanceTest extends DistributionTest {
 
         private void addExecutionParameters(List<String> result) {
             addSystemPropertyIfExist(result, "org.gradle.performance.scenarios", scenarios)
+            addSystemPropertyIfExist(result, "org.gradle.performance.testProject", getTestProjectName().getOrNull())
```
Comment (lptr) [inline buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy:274]: Can this ever be `null`?

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt [hunk 4]
```diff
@@ -244,6 +245,44 @@ class PerformanceTestPlugin : Plugin<Project> {
...
+                val channelSuffix = if (OperatingSystem.current().isLinux) "" else "-${OperatingSystem.current().familyName.toLowerCase()}"
+                create("${sampleGenerator.name}PerformanceTest") {
+                    performanceReporter = createPerformanceReporter()
+                    channel = "commits$channelSuffix"
+                    setTestProjectGenerationTask(sampleGenerator)
+
+                    if (shouldLoadScenariosFromFile) {
+                        scenarios = loadScenariosFromFile(sampleGenerator.name).joinToString(";")
+                        doFirst {
+                            assert(scenarios.isNotEmpty()) { "Running $name requires to set some scenarios" }
```
Comment (lptr) [inline buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:271]: Does `$name` here refer to `sampleGenerator.name`?

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt [hunk 5]
```diff
@@ -244,6 +245,44 @@ class PerformanceTestPlugin : Plugin<Project> {
...
             outputs.doNotCacheIf("Is adhoc performance test") { true }
         }
+
+        val shouldLoadScenariosFromFile = project.providers.gradleProperty("includePerformanceTestScenarios")
+            .forUseAtConfigurationTime()
+            .getOrElse("false") == "true"
+
+        configureSampleGenerators {
+            all {
+                val sampleGenerator = this
```
Comment (lptr) [inline buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:255]: Why don't we do something like this instead?

```suggestion
            all { sampleGenerator ->
```

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt [hunk 6]
```diff
@@ -244,6 +245,44 @@ class PerformanceTestPlugin : Plugin<Project> {
...
+                create("${sampleGenerator.name}PerformanceAdHocTest") {
+                    performanceReporter = createPerformanceReporter()
+                    channel = "adhoc"
+                    outputs.doNotCacheIf("Is adhoc performance test") { true }
+                    setTestProjectGenerationTask(sampleGenerator)
+                }
+                val channelSuffix = if (OperatingSystem.current().isLinux) "" else "-${OperatingSystem.current().familyName.toLowerCase()}"
+                create("${sampleGenerator.name}PerformanceTest") {
```
Comment (lptr) [inline buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:263]: Would be great to add descriptions to these tasks.

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt [hunk 7]
```diff
@@ -244,6 +245,44 @@ class PerformanceTestPlugin : Plugin<Project> {
             channel = "adhoc"
             outputs.doNotCacheIf("Is adhoc performance test") { true }
         }
+
+        val shouldLoadScenariosFromFile = project.providers.gradleProperty("includePerformanceTestScenarios")
+            .forUseAtConfigurationTime()
+            .getOrElse("false") == "true"
+
+        configureSampleGenerators {
```
Comment (lptr) [inline buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:253]: So "sample" and "test project" is the same, right? Can we standardize on one or the other?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/CrossVersionPerformanceTestRunner.groovy [hunk 8]
```diff
@@ -106,6 +108,13 @@ class CrossVersionPerformanceTestRunner extends PerformanceTestSpec {
         return measuredBuildOperations
     }
 
+    void loadConfiguredTestProject() {
+        def testProjectFromSystemProperty = System.getProperty(TEST_PROJECT_PROPERTY_NAME)
+        if (testProjectFromSystemProperty != null && !testProjectFromSystemProperty.empty) {
+            testProject = testProjectFromSystemProperty
+        }
+    }
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/CrossVersionPerformanceTestRunner.groovy:116]: Given that this method is called from the constructor, can we make it `static` and let it return the state to be set?

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaUpToDatePerformanceTest.groovy [hunk 9]
```diff
@@ -69,9 +62,14 @@ class JavaUpToDatePerformanceTest extends AbstractCrossVersionPerformanceTest {
...
+        parallel << [true, false]
+    }
+
+    private JavaTestProject getJavaTestProject() {
+        def javaTestProject = JavaTestProject.values().find { it.projectName == runner.testProject }
+        if (javaTestProject == null) {
+            throw new IllegalArgumentException("Cannot find Java test project for ${runner.testProject}")
+        }
+        return javaTestProject
     }
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaUpToDatePerformanceTest.groovy:74]: Hm, why is this code here? Won't it be used for every performance test?

File: .teamcity/Gradle_Check/model/FunctionalTestBucketProvider.kt (hunk 1 above)
Comment (wolfs) [reply .teamcity/Gradle_Check/model/FunctionalTestBucketProvider.kt:183]: Idea tells me that `number` is never used as a `val`, so I removed the `val` making the property only available while constructing the object.

File: buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy (hunk 3 above)
Comment (wolfs) [reply buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy:274]: It is currently `null` for all the old performance test tasks. When we are finished migrating to the new infrastructure, it should never be null.

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt (hunk 4 above)
Comment (wolfs) [reply buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:271]: No, it refers to the task name.

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt (hunk 5 above)
Comment (wolfs) [reply buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:255]: This would call the wrong `all` method - the one from Kotlin returning a `boolean`.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaUpToDatePerformanceTest.groovy (hunk 9 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaUpToDatePerformanceTest.groovy:74]: Yeah, this needs to go somewhere else. Currently, this is the only performance test converted to the new infrastructure, this is why I left it here.

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt (hunk 4 above)
Comment (lptr) [reply buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:271]: What does `sampleGenerator.name` refer to?

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt (hunk 5 above)
Comment (lptr) [reply buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:255]: This is very confusing. :(

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt (hunk 4 above)
Comment (wolfs) [reply buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:271]: Let's do an example:
* `sampleGenerator.name == largeJavaMultiProject`
* `name == largeJavaMultiProjectPerformanceTest`.

**Review by blindpirate — COMMENTED** [review]
What do you think if we can make every scenario a bucket? If so, instead of 

- Performance test for largeJavaMultiProject (bucket 1)
- Performance test for largeJavaMultiProject (bucket 2)
- ...

We can have:

- clean assemble on XXX (parallel true)
- clean assemble on XXX (parallel false)
- ...

Given that test scenarios are not changing a lot. If we use 40 buckets, there'll be 2~3 tests in a bucket, that might have bad impact on feedback time if we have enough machines?

**Comment by wolfs** [comment]
> What do you think if we can make every scenario a bucket?

I am not opposed to that. Though I actually think that the feedback time will improve when we use fewer builds. Let's try this out.

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt (hunk 5 above)
Comment (wolfs) [reply buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:274]: It is, and I got it wrong and didn't understand what is going on.

File: buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy [hunk 10]
```diff
@@ -49,10 +55,7 @@ import java.nio.charset.Charset
 abstract class PerformanceTest extends DistributionTest {
     public static final String TC_URL = "https://builds.gradle.org/viewLog.html?buildId="
     public static final Set<String> NON_CACHEABLE_VERSIONS = Sets.newHashSet("last", "nightly", "flakiness-detection-commit")
-    // Baselines configured by command line `--baselines`
-    private Property<String> configuredBaselines = getProject().getObjects().property(String.class)
-    // Baselines determined by determineBaselines task
-    private Property<String> determinedBaselines = getProject().getObjects().property(String.class)
+    private final Property<String> baselines = getProject().getObjects().property(String.class)
```
Comment (lptr) [inline buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy:58]: Is this a comma-separated list? We should say so in Javadoc.

Also, can we use an abstract class with getters for the properties?

Comment (lptr) [reply buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy:58]: Also, would it make things worse if we used a `ListProperty` here?

I think it's fine to follow up on any of this in a separate PR.

File: buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy [hunk 11]
```diff
@@ -117,38 +120,56 @@ abstract class PerformanceTest extends DistributionTest {
...
+                    scenarios.add(methodName)
+                    return className
+                } else {
+                    return includePattern
+                }
+            }
+        } as Set
+        filter.setCommandLineIncludePatterns(onlyClassFilters)
+        setScenarios(scenarios.join(";"))
+    }
```
Comment (lptr) [inline buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy:162]: This code is weird. We have a `collect()` call that also has side-effects. Then we mutate the inputs inside the task action. Can we straighten this out?

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt (hunk 5 above)
Comment (lptr) [reply buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:274]: @eskatos @bamboo can you guys give us a hand here? How should this code be written to reduce confusion? Thanks!

File: buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy (hunk 11 above)
Comment (wolfs) [reply buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy:162]: I'll remove the side-effect from the `collect` call.

This is a workaround for test filters not working for unrolled spock tests. What I can do to make it work better is to remove the whole `scenarios` as an input, and only allow folks to use `--tests`. Then the above mutation of the inputs is `benign`, in the sense that it leaves the test filter in place and replaces it by something that works.

We have a fix for filtering unrolled Spock tests on `master`, though we didn't switch to the JUnit 5 platform, yet. As soon as we do that, the whole workaround can go away.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjects.groovy [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+package org.gradle.performance.generator
+
+class TestProjects {
+    static List<String> getProjectMemoryOptions(String testProject) {
+        def daemonMemory = determineDaemonMemory(testProject)
+        return ["-Xms${daemonMemory}", "-Xmx${daemonMemory}"]
+    }
+
+    private static String determineDaemonMemory(String testProject) {
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjects.groovy:25]: Why do it like this? Can we embed this information in `JavaTestProject` instead?

Can we create an enum of sorts for these test projects?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/GradleProfilerReporter.java [hunk 13]
```diff
@@ -27,17 +27,14 @@
...
 public class GradleProfilerReporter implements DataReporter<PerformanceTestResult> {
     private static final String DEBUG_ARTIFACTS_DIRECTORY_PROPERTY_NAME = "org.gradle.performance.debugArtifactsDirectory";
-    private final BenchmarkResultCollector resultCollector;
+    private BenchmarkResultCollector resultCollector;
+    private final File debugArtifactsDirectory;
 
     public GradleProfilerReporter(File fallbackDirectory) {
         String debugArtifactsDirectoryPath = System.getProperty(DEBUG_ARTIFACTS_DIRECTORY_PROPERTY_NAME);
-        File debugArtifactsDirectory = Strings.isNullOrEmpty(debugArtifactsDirectoryPath)
+        debugArtifactsDirectory = Strings.isNullOrEmpty(debugArtifactsDirectoryPath)
```
Comment (lptr) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/GradleProfilerReporter.java:35]:
```suggestion
        this.debugArtifactsDirectory = Strings.isNullOrEmpty(debugArtifactsDirectoryPath)
```

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingNativePerformanceTest.groovy [hunk 14]
```diff
@@ -17,11 +17,8 @@
 package org.gradle.performance.regression.buildcache
 
 import org.gradle.initialization.ParallelismBuildOptions
-import org.gradle.util.Requires
-import org.gradle.util.TestPrecondition
-import spock.lang.Unroll
 
-@Requires(TestPrecondition.LINUX)
+// TODO: Merge with TaskOutputCachingJavaPerformanceTest
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingNativePerformanceTest.groovy:21]: When shall we do this? Should there be an issue open for this?

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingSwiftPerformanceTest.groovy [hunk 15]
```diff
@@ -19,32 +19,24 @@ package org.gradle.performance.regression.buildcache
 import org.gradle.initialization.ParallelismBuildOptions
 import org.gradle.util.Requires
 import org.gradle.util.TestPrecondition
-import spock.lang.Unroll
 
-@Requires(TestPrecondition.LINUX)
+@Requires(TestPrecondition.NOT_WINDOWS)
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingSwiftPerformanceTest.groovy:23]: I think it's better to have an exhaustive list of OS where we do want to run these things. WDYT?

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/corefeature/TaskCreationPerformanceTest.groovy [hunk 16]
```diff
@@ -22,7 +22,6 @@ class TaskCreationPerformanceTest extends AbstractCrossVersionPerformanceTest {
 
     def "create many tasks"() {
         given:
-        runner.testProject = "createLotsOfTasks"
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/corefeature/TaskCreationPerformanceTest.groovy:25]: Does running this on any other test project make sense?

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingSwiftPerformanceTest.groovy (hunk 15 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingSwiftPerformanceTest.groovy:23]: Yes, I think the whole `@Requires` annotations should go away.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaCleanAssemblePerformanceTest.groovy [hunk 17]
```diff
@@ -17,37 +17,22 @@
...
-        runner.gradleOpts = ["-Xms${testProject.daemonMemory}", "-Xmx${testProject.daemonMemory}"]
-        runner.warmUpRuns = warmUpRuns
-        runner.runs = runs
+        runner.gradleOpts = runner.projectMemoryOptions
+        runner.warmUpRuns = 2
+        runner.runs = 6
         runner.tasksToRun = ["clean", "assemble"]
         runner.targetVersions = ["6.7-20200824220048+0000"]
-        runner.minimumBaseVersion = minimumBaseVersion
+        runner.minimumBaseVersion = runner.testProject.contains("Composite") ? "4.0" : null
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaCleanAssemblePerformanceTest.groovy:30]: This sounds like something that would be better modeled via something like `testProject.requiresComposite`, or maybe set automatically from `testProject.minimumBaseVersion`.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaConfigurationCachePerformanceTest.groovy [hunk 18]
```diff
@@ -83,16 +80,13 @@ class JavaConfigurationCachePerformanceTest extends AbstractCrossVersionPerforma
         then:
         result.assertCurrentVersionHasNotRegressed()
 
+
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaConfigurationCachePerformanceTest.groovy:83]: Nit: unnecessary newline.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/corefeature/TaskCreationPerformanceTest.groovy (hunk 16 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/corefeature/TaskCreationPerformanceTest.groovy:25]: No. Actually, we should merge this test with `JavaConfigurationPerformanceTest` and add `createLotsOfTasks ` as a test project there, WDYT?

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaIDEModelPerformanceTest.groovy [hunk 19]
```diff
@@ -138,20 +129,20 @@ class JavaIDEModelPerformanceTest extends AbstractCrossVersionPerformanceTest {
...
 
-    private setupTestProject(JavaTestProject testProject, int iterations) {
-        runner.testProject = testProject
-        runner.gradleOpts = ["-Xms${testProject.daemonMemory}", "-Xmx${testProject.daemonMemory}"]
+    private setupTestProject() {
+        runner.gradleOpts = runner.projectMemoryOptions
+        def iterations = determineIterations()
         runner.warmUpRuns = iterations
         runner.runs = iterations
     }
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaIDEModelPerformanceTest.groovy:139]: This sets up the runner, or the iterations, not the test project.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/NativeBuildDependentsPerformanceTest.groovy [hunk 20]
```diff
@@ -30,12 +30,16 @@ class NativeBuildDependentsPerformanceTest extends AbstractCrossVersionPerforman
...
+        // 'largeNativeBuild'     | 'project432:buildDependentsExternalComponent111'
+        // TODO Re-evaluate this scenario: memory consumption stress case, gradleOpts = ['-Xms4g', '-Xmx4g']
+        // The generated dependency graph is rather complex and deep, unrealistic?
+        // 'nativeDependentsDeep' | 'libA0:buildDependentsLibA0'
         given:
-        runner.testProject = testProject
         runner.tasksToRun = [ task ]
         runner.args += ["--parallel", "--max-workers=4"]
-        runner.gradleOpts = ["-Xms3g", "-Xmx3g"]
+        runner.gradleOpts = runner.projectMemoryOptions
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/NativeBuildDependentsPerformanceTest.groovy:42]: Shouldn't this be set automatically?

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/RealWorldNativePluginPerformanceTest.groovy [hunk 21]
```diff
@@ -110,13 +104,33 @@ class RealWorldNativePluginPerformanceTest extends AbstractCrossVersionPerforman
...
+            case 'source':
+                return 'modules/project5/src/src100_c.c'
+            case 'header':
+                return 'modules/project1/src/src50_h.h'
+            case 'build':
+                return 'common.gradle'
+            default:
+                throw new IllegalArgumentException("Unknown change type ${changeType}")
+        }
+    }
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/RealWorldNativePluginPerformanceTest.groovy:121]: Let's have an enum for `changeType`.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/SwiftBuildPerformanceTest.groovy [hunk 22]
```diff
@@ -22,53 +22,49 @@ import org.gradle.profiler.BuildContext
...
+    static String determineFileToChange(String testProject) {
+        switch (testProject) {
+            case 'mediumSwiftMulti':
+                return 'lib6api3/src/main/swift/Lib6Api3Impl2Api.swift'
+            case 'bigSwiftApp':
+                return 'src/main/swift//AppImpl54Api3.swift'
+            default:
+                throw new IllegalArgumentException("Invalid test project ${testProject}")
+        }
     }
```
Comment (lptr) [inline subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/SwiftBuildPerformanceTest.groovy:68]: I wonder if we could capture this in some better way, like declaring the test project as `SwiftTestProject` that has a property called `fileToChange` or something. Should `IncrementalTestProject` already have such a property BTW?

Comment (lptr) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/SwiftBuildPerformanceTest.groovy:68]: Also, it would be nice to declaratively capture that this test requires a `SwiftTestProject` and cannot be run with anything else.

But if we model this via an `IncrementalTestProject`, can't we just have this work the same way we have with incremental Java compile? And maybe Groovy?

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt (hunk 5 above)
Comment (eskatos) [reply buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt:274]: Either you capture `this` in an explicit variable as done above, or you use the implicit receiver directly. There isn't more to it.

File: .teamcity/Gradle_Check/model/bucket-extensions.kt (hunk 2 above)
Comment (wolfs) [reply .teamcity/Gradle_Check/model/bucket-extensions.kt]: I changed the name to `splitIntoBuckets` and added separate lines for each parameter.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/CrossVersionPerformanceTestRunner.groovy (hunk 8 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/CrossVersionPerformanceTestRunner.groovy]: Done.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaUpToDatePerformanceTest.groovy (hunk 9 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaUpToDatePerformanceTest.groovy]: This is now in `TestProjects`.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjects.groovy (hunk 12 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjects.groovy:25]: We can, I opened https://github.com/gradle/gradle/issues/14597 to make this better.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingNativePerformanceTest.groovy (hunk 14 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingNativePerformanceTest.groovy:21]: I created https://github.com/gradle/gradle/issues/14598.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/NativeBuildDependentsPerformanceTest.groovy (hunk 20 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/NativeBuildDependentsPerformanceTest.groovy:42]: I am not sure where to set this. I actually think the performance test project should define the memory settings in its `gradle.properties`. And the performance test should fail if the test project fails to declare anything. WDYT?

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/SwiftBuildPerformanceTest.groovy (hunk 22 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/nativeplatform/SwiftBuildPerformanceTest.groovy:68]: Yes. We should handle this as part of https://github.com/gradle/gradle/issues/14597.

File: buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt (hunk 7 above)
Comment (wolfs) [reply buildSrc/subprojects/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt]: Changed the method to `configureSampleGenerators()`.

File: buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy (hunk 11 above)
Comment (wolfs) [reply buildSrc/subprojects/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy:170]: I changed the collect call to `each` and added some documentation why we do this.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/GradleProfilerReporter.java (hunk 13 above)
Comment (wolfs) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/GradleProfilerReporter.java]: Fixed.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingSwiftPerformanceTest.groovy (hunk 15 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/buildcache/TaskOutputCachingSwiftPerformanceTest.groovy]: I removed the `@Requires` annotations.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaCleanAssemblePerformanceTest.groovy (hunk 17 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaCleanAssemblePerformanceTest.groovy:30]: This should be handled as part of https://github.com/gradle/gradle/issues/14597 as well I think.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaConfigurationCachePerformanceTest.groovy (hunk 18 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaConfigurationCachePerformanceTest.groovy]: Removed.

File: subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaIDEModelPerformanceTest.groovy (hunk 19 above)
Comment (wolfs) [reply subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/java/JavaIDEModelPerformanceTest.groovy:139]: Renamed to `setupRunner()`.
